apps/studio/state/storage-explorer.tsx

```diff
@@ -1417,6 +1532,406 @@ function createStorageExplorerState({
+          toast.error(`Failed to move folder ${folder.name}: ${error.message}`)
+        }
+      }
+
+      // Handle regular file moves
+      await Promise.all(
+        files.map(async (item) => {
+          const pathToFile = state.openedFolders
+            .slice(0, item.columnIndex)
+            .map((folder) => folder.name)
+            .join('/')
+
+          const fromPath = pathToFile.length > 0 ? `${pathToFile}/${item.name}` : item.name
+          const toPath =
+            newPathToFile.length > 0 ? `${formattedNewPathToFile}/${item.name}` : item.name
+
+          try {
+            await moveStorageObject({
+              projectRef: state.projectRef,
+              bucketId: state.selectedBucket.id,
+              from: fromPath,
+              to: toPath,
+            })
+          } catch (error: any) {
+            numberOfFilesMovedFail += 1
+            toast.error(error.message)
+          }
+        })
+      )
+
+      if (numberOfFilesMovedFail === items.length) {
+        toast.error('Failed to move items')
+      } else {
+        const movedFiles = items.length - numberOfFilesMovedFail
+        toast(
+          `Successfully moved ${movedFiles} item${movedFiles > 1 ? 's' : ''} to ${formattedNewPathToFile.length > 0 ? formattedNewPathToFile : 'the root of your bucket'}`
+        )
+      }
+
+      toast.dismiss(toastId)
+
+      // Clear selected items after successful move operation
+      state.clearSelectedItems()
+
+      // TODO: invalidate the file preview cache when moving files
+      await state.validateAndUpdateOpenedFolders()
+    },
+
+    // New function to handle folder moves by creating new folder and moving contents
+    moveFolderWithContents: async (folder: StorageItemWithColumn, targetPath: string) => {
+      const folderName = folder.name
+      const sourcePath = state.openedFolders
+        .slice(0, folder.columnIndex)
+        .map((folder) => folder.name)
+        .join('/')
+
+      const sourceFullPath = sourcePath.length > 0 ? `${sourcePath}/${folderName}` : folderName
+      const targetFullPath = targetPath.length > 0 ? `${targetPath}/${folderName}` : folderName
+
+      // Add folder to the "being moved" set for visual feedback
+      state.addFolderBeingMoved(targetFullPath)
+
+      // Create a placeholder file to create the new folder
+      const placeholderContent = new Blob([''], { type: 'text/plain' })
+
+      try {
+        // Check if supabaseClient is available
+        if (!state.supabaseClient) {
+          throw new Error('Supabase Client is not available')
+        }
+
+        // Check if selectedBucket and projectRef are available
+        if (!state.selectedBucket || !state.selectedBucket.id) {
+          throw new Error('Selected bucket is not available')
+        }
+
+        if (!state.projectRef) {
+          throw new Error('Project reference is not available')
+        }
+
+        // Create the new folder by uploading a placeholder file
```

Comment:
we don't need to do this actually - folders are an FE concept
so if there's a path say '/path1/path2/file.tsx` then path 1 and path 2 are naturally folders